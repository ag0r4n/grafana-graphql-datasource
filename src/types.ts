import {DataQuery, DataSourceJsonData} from '@grafana/ui';

export interface GraphQL {
  query: string;
  variables?: any;
}

export interface GraphQLQuery extends DataQuery {
  request: GraphQL;
  jsonata?: string;
}

export interface GraphQLOptions extends DataSourceJsonData {
  authHeader?: string;
}
