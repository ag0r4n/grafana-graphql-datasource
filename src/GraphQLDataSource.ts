// Types
import {
  DataQueryRequest,
  DataQueryResponse,
  DataSourceApi,
  DataSourceInstanceSettings,
} from '@grafana/ui';

import {
  DataFrame,
  isDataFrame,
  toDataFrame,
} from '@grafana/data';

import truncate from 'lodash/truncate';

import {GraphQLQuery, GraphQLOptions} from './types';
import jsonata from 'jsonata';

export class GraphQLDataSource extends DataSourceApi<GraphQLQuery, GraphQLOptions> {
  constructor(private instanceSettings: DataSourceInstanceSettings<GraphQLOptions>) {
    super(instanceSettings);
  }

  getQueryDisplayText(query: GraphQLQuery) {
    if (!query.request || !query.request.query) {
      return 'Missing Query';
    }
    return truncate(query.request.query, {length: 20});
  }

  query(options: DataQueryRequest<GraphQLQuery>): Promise<DataQueryResponse> {
    const all = options.targets.map(q => this.doSingleQuery(q));
    return Promise.all(all).then(data => {
      return {data};
    });
  }

  async doSingleQuery(query: GraphQLQuery): Promise<DataFrame> {
    const opts = {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(query.request),
    };
    // Add Authorization header
    if (this.instanceSettings.jsonData.authHeader) {
      (opts.headers as any).Authorization = this.instanceSettings.jsonData.authHeader;
    }

    const res = await fetch(this.instanceSettings.url!, opts);
    let json = await res.json();
    if (query.jsonata) {
      json = jsonata(query.jsonata).evaluate(json);
    }

    // Make sure it looks like DataFrame
    if (isDataFrame(json)) {
      return json as DataFrame;
    }
    return toDataFrame(json);
  }

  testDatasource() {
    const url = this.instanceSettings.url;
    if (!url) {
      return Promise.resolve({
        status: 'warn',
        message: 'Missing URL',
      });
    }

    return fetch(url, {
      method: 'GET',
    }).then((response: any) => {
      console.log('RESPONSE', response);
      return {
        status: 'success',
        message: 'OK',
      };
    });
  }
}
