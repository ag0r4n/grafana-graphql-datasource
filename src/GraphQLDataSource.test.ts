import { GraphQLDataSource } from './GraphQLDataSource';
import { GraphQLOptions, GraphQLQuery } from './types';
import { DataSourceInstanceSettings, PluginMeta, DataQueryRequest } from '@grafana/ui';

describe('GraphQLDatasource', () => {
  const instanceSettings: DataSourceInstanceSettings<GraphQLOptions> = {
    id: 1,
    type: 'x',
    name: 'xxx',
    url: 'hello',
    meta: {} as PluginMeta,
    jsonData: {},
  };

  describe('when querying', () => {
    test('should give error with no query defined', () => {
      const ds = new GraphQLDataSource(instanceSettings);
      const query = {
        refId: 'Z',
      } as GraphQLQuery;

      const options = {
        targets: [query],
      } as DataQueryRequest<GraphQLQuery>;

      // Missing path
      return ds
        .query(options)
        .then(rsp => {
          expect(true).toBeFalsy(); // FAIL!
        })
        .catch(reason => {
          expect(reason).toMatchSnapshot();
        });
    });
  });
});
