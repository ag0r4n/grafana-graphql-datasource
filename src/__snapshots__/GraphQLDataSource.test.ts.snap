// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQLDatasource when querying should give error with no query defined 1`] = `[ReferenceError: fetch is not defined]`;
